### ISSUE TITLE
Edge case with `no-unused-expressions` rule

### ISSUE THREAD

**Issue by markelog** [eslint/eslint#6636]
**What version of ESLint are you using?**
3.0.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

``` json
{   
    "rules": {
        "no-unused-expressions": "error"
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

``` js
one = 1,
two = 2
```

**What did you expect to happen?**
To not show the error

**What actually happened? Please include the actual, raw output from ESLint.**
it shows the error for the first line

``` sh
1:1  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions
```

---

It looks like an edge case for not really good code (http://eslint.org/docs/rules/no-sequences), but seems this is still an error

**Comment by btmills** [comment]
This is behaving as I would expect it to. It's not complaining about the assignment expressions, just the sequence expression. If you get rid of the sequence expression by removing the comma, then it shouldn't complain anymore, and the code will still be functionally identical.

**Comment by platinumazure** [comment]
Well, is there perhaps room for an option where, given a SequenceExpression, the rule would check that all elements of the sequence have side effects and warn if any do not?

**Comment by markelog** [comment]
> If you get rid of the sequence expression by removing the comma, then it shouldn't complain anymore, and the code will still be functionally identical.

Yes, that code is identical i.e. all expression is such code are used, which looks like contradiction to the meaning of that rule?

**Comment by markelog** [comment]
To get things a bit more interesting, this code also will be flagged - 
```js
console.log(1), console.log(2)
```

**Comment by btmills** [comment]
The expressions inside the sequence both have side effects, so they're fine, but the sequence expression itself is unused and unnecessary.

**Comment by platinumazure** [comment]
@btmills What about expressions like `a() && b()`? Couldn't we say the conditional expression could also be useless even though each part of the conditional expression is useless?

**Comment by btmills** [comment]
In that case, the `CallExpression`s `a()` and `b()` would be useful as they may have side effects, and the `LogicalExpression` should be useful due to the short-circuiting behavior, in which it would execute `b()` if and only if `a()` returns a truthy value.

**Comment by markelog** [comment]
> sequence expression itself is unused

Sorry i don't understand this train of thought

>  and unnecessary.

Yes, that statement is clear and real to me but we have different rule for it - http://eslint.org/docs/rules/no-sequences

**Comment by btmills** [comment]
The [no-unused-expressions docs](http://eslint.org/docs/rules/no-unused-expressions) cover this case:

> Sequence expressions (those using a comma, such as a = 1, b = 2) are always considered unused unless their return value is assigned or used in a condition evaluation, or a function call is made with the sequence expression value.

Based on that, a sequence expression is unused unless the return value is used.

**Comment by markelog** [comment]
This is still peculiar logic to me but if discussion was already transpired on it i'm closing this

**Comment by markelog** [comment]
btw, what are correct labels for ticket like this?

**Comment by platinumazure** [comment]
Anything that turns out not to be a bug or enhancement request is usually labeled "question".

Personally I think there's something to be said for at least having a better freaking message for this case, because it's confusing as hell right now.

I'd like to reopen as an enhancement request to make the error message for SequenceExpression more clear (maybe even say, did you mean to have a semicolon between the expressions). @markelog @btmills Either of you object to that?

**Comment by btmills** [comment]
Something like "A sequence expression is unnecessary when its return value is unused"?

**Comment by platinumazure** [comment]
@btmills Sure, just something to clue people into the notion that it's the sequence expression that's the problem, not either of the subexpressions. That would work for me.

**Comment by btmills** [comment]
As long as there are no compatibility issues with changing a message in a minor release, :+1: from me.

**Comment by platinumazure** [comment]
Then we need a champion and two :+1: (or one if we count @markelog), I think.

**Comment by markelog** [comment]
Some people consider changes in output as breaking ones but i think it would be a big stretch to consider this only for major since breaking changes is more about output format not about info inside.

I think changing message would be a good idea (although i think separate issue would be better, since this one is not about the message, but it is a bit bureaucratic to do so) even though this is still a bizarre and confusing for me either way.

I also don't think message should have "unnecessary" word in there, cause this rule about "unused" expressions and i'm assuming we are consider comma as "unused" since what is "unnecessary" should be out of scope of this rule

**Comment by michaelficarra** [comment]
👎 There are so many expressions that are not assignments or function calls. I don't think a custom message is needed for each of them, and I especially don't think a custom message is needed for sequences since they are so uncommon.

**Comment by kaicataldo** [comment]
Doesn't look like we've made headway towards consensus on this in a while. Do we want to keep the discussion going or close this out?

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
